File: streampark-common/src/main/scala/org/apache/streampark/common/util/FileUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.common.util

import java.io._
import java.net.URL
import java.util
import java.util.Scanner

import scala.collection.JavaConversions._
import scala.collection.mutable

object FileUtils {

  private[this] def bytesToHexString(src: Array[Byte]): String = {
    val stringBuilder = new mutable.StringBuilder
    if (src == null || src.length <= 0) return null
    for (i <- src.indices) {
      val v: Int = src(i) & 0xff
      val hv: String = Integer.toHexString(v).toUpperCase
      if (hv.length < 2) {
        stringBuilder.append(0)
      }
      stringBuilder.append(hv)
    }
    stringBuilder.toString
  }

  def isJarFileType(input: InputStream): Boolean = {
    if (input == null) {
      throw new RuntimeException("The inputStream can not be null")
    }
    Utils.using(input) {
      in =>
        val b = new Array[Byte](4)
        in.read(b, 0, b.length)
        bytesToHexString(b)
    } == "504B0304"
  }

  def isJarFileType(file: File): Boolean = {
    if (!file.exists || !file.isFile) {
      throw new RuntimeException("The file does not exist or the path is a directory")
    }
    isJarFileType(new FileInputStream(file))
  }

  def createTempDir(): File = {
    val TEMP_DIR_ATTEMPTS = 10000
    val baseDir = new File(System.getProperty("java.io.tmpdir"))
    val baseName = System.currentTimeMillis + "-"
    for (counter <- 0 until TEMP_DIR_ATTEMPTS) {
      val tempDir = new File(baseDir, baseName + counter)
      if (tempDir.mkdir) {
        return tempDir
      }
    }
    throw new IllegalStateException(
      s"[StreamPark] Failed to create directory within $TEMP_DIR_ATTEMPTS  attempts (tried $baseName 0 to $baseName ${TEMP_DIR_ATTEMPTS - 1})")
  }

  def exists(path: String): Unit = {
    require(
      path != null && path.nonEmpty && new File(path).exists(),
      s"[StreamPark] FileUtils.exists: file $path is not exist!")
  }

  def mkdir(dir: File) = {
    if (dir.exists && !dir.isDirectory) {
      throw new IOException(s"File $dir exists and is not a directory. Unable to create directory.")
    } else if (!dir.mkdirs) {
      // Double-check that some other thread or process hasn't made
      if (!dir.isDirectory) {
        throw new IOException(s"Unable to create directory $dir")
      }
    }
  }

  def getPathFromEnv(env: String): String = {
    val path = System.getenv(env)
    require(
      Utils.notEmpty(path),
      s"[StreamPark] FileUtils.getPathFromEnv: $env is not set on system env")
    val file = new File(path)
    require(file.exists(), s"[StreamPark] FileUtils.getPathFromEnv: $env is not exist!")
    file.getAbsolutePath
  }

  def resolvePath(parent: String, child: String): String = {
    val file = new File(parent, child)
    require(
      file.exists,
      s"[StreamPark] FileUtils.resolvePath: ${file.getAbsolutePath} is not exist!")
    file.getAbsolutePath
  }

  def getSuffix(filename: String): String = {
    require(filename != null)
    filename.drop(filename.lastIndexOf("."))
  }

  def listFileAsURL(dirPath: String): util.List[URL] = {
    new File(dirPath) match {
      case x if x.exists() && x.isDirectory =>
        val files = x.listFiles()
        if (files != null && files.nonEmpty) {
          files.map(f => f.toURI.toURL).toList
        } else {
          util.Collections.emptyList()
        }
      case _ => util.Collections.emptyList()
    }
  }

  def equals(file1: File, file2: File): Boolean = {
    (file1, file2) match {
      case (a, b) if a == null || b == null => false
      case (a, b) if !a.exists() || !b.exists() => false
      case (a, b) if a.getAbsolutePath == b.getAbsolutePath => true
      case (a, b) =>
        val first = new BufferedInputStream(new FileInputStream(a))
        val second = new BufferedInputStream(new FileInputStream(b))
        if (first.available() != second.available()) false;
        else {
          while (true) {
            val firRead = first.read()
            val secRead = second.read()
            if (firRead != secRead) {
              Utils.close(first, second)
              return false
            }
            if (firRead == -1) {
              Utils.close(first, second)
              return true;
            }
          }
          true
        }
    }
  }

  @throws[IOException]
  def readString(file: File): String = {
    require(file != null && file.isFile)
    val reader = new FileReader(file)
    val scanner = new Scanner(reader)
    val buffer = new mutable.StringBuilder()
    if (scanner.hasNextLine) {
      buffer.append(scanner.nextLine())
    }
    while (scanner.hasNextLine) {
      buffer.append("\r\n")
      buffer.append(scanner.nextLine())
    }
    Utils.close(scanner, reader)
    buffer.toString()
  }

  @throws[IOException]
  def readString(in: InputStream): String = {
    require(in != null)
    val scanner = new Scanner(in)
    val buffer = new mutable.StringBuilder()
    if (scanner.hasNextLine) {
      buffer.append(scanner.nextLine())
    }
    while (scanner.hasNextLine) {
      buffer.append("\r\n")
      buffer.append(scanner.nextLine())
    }
    Utils.close(scanner)
    buffer.toString()
  }

}

File: streampark-common/src/main/java/org/apache/streampark/common/util/FileUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.io.Serializable;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public final class FileUtils {

    private FileUtils() {
    }

    public static File toCanonicalFile(String path) throws IOException {
        Path normalized = Paths.get(path).normalize();
        if (normalized.toString().contains("..")) {
            throw new IOException("Invalid path: " + path);
        }
        return normalized.toFile().getCanonicalFile();
    }

    public static File resolveChildFile(File baseDir, String... childSegments) throws IOException {
        Path base = baseDir.getCanonicalFile().toPath();
        Path current = base;
        for (String segment : childSegments) {
            if (segment == null || segment.isEmpty() || segment.contains("..")) {
                throw new IOException("Invalid path segment: " + segment);
            }
            current = current.resolve(segment);
        }
        Path resolved = current.normalize().toAbsolutePath();
        if (!resolved.startsWith(base)) {
            throw new IOException("Resolved path escapes base directory: " + resolved);
        }
        return resolved.toFile();
    }

    private static String bytesToHexString(byte[] src) {
        if (src == null || src.length <= 0) {
            return null;
        }
        StringBuilder stringBuilder = new StringBuilder();
        for (byte b : src) {
            int v = b & 0xFF;
            String hv = Integer.toHexString(v).toUpperCase();
            if (hv.length() < 2) {
                stringBuilder.append(0);
            }
            stringBuilder.append(hv);
        }
        return stringBuilder.toString();
    }

    public static boolean isJarFileType(InputStream input) {
        if (input == null) {
            throw new RuntimeException("The inputStream can not be null");
        }
        return AutoCloseUtils.using(
            input,
            in -> {
                byte[] b = new byte[4];
                try {
                    in.read(b, 0, b.length);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
                return bytesToHexString(b);
            })
            .equals("504B0304");
    }

    public static boolean isJarFileType(File file) throws IOException {
        if (!file.exists() || !file.isFile()) {
            throw new RuntimeException("The file does not exist or the path is a directory");
        }
        return isJarFileType(new FileInputStream(file));
    }

    public static File createTempDir() {
        final int TEMP_DIR_ATTEMPTS = 10000;
        File baseDir = new File(System.getProperty("java.io.tmpdir"));
        String baseName = System.currentTimeMillis() + "-";
        for (int counter = 0; counter < TEMP_DIR_ATTEMPTS; counter++) {
            File tempDir = new File(baseDir, baseName + counter);
            if (tempDir.mkdir()) {
                return tempDir;
            }
        }
        throw new IllegalStateException(
            "[StreamPark] Failed to create directory within "
                + TEMP_DIR_ATTEMPTS
                + "  attempts (tried "
                + baseName
                + " 0 to "
                + baseName
                + (TEMP_DIR_ATTEMPTS - 1)
                + ")");
    }

    public static void mkdir(File dir) throws IOException {
        if (dir.exists() && !dir.isDirectory()) {
            throw new IOException("File " + dir + " exists and is not a directory. Unable to create directory.");
        } else if (!dir.mkdirs() && !dir.isDirectory()) {
            throw new IOException("Unable to create directory " + dir);
        }
    }

    public static String getPathFromEnv(String env) {
        String path = System.getenv(env);
        if (path == null) {
            path = System.getProperty(env);
        }
        AssertUtils.notNull(path, "[StreamPark] FileUtils.getPathFromEnv: " + env + " is not set on system env");
        File file = new File(path);
        if (!file.exists()) {
            throw new IllegalArgumentException("[StreamPark] FileUtils.getPathFromEnv: " + env + " is not exist!");
        }
        return file.getAbsolutePath();
    }

    public static String resolvePath(String parent, String child) {
        File file = new File(parent, child);
        if (!file.exists()) {
            throw new IllegalArgumentException(
                "[StreamPark] FileUtils.resolvePath: " + file.getAbsolutePath() + " is not exist!");
        }
        return file.getAbsolutePath();
    }

    public static String getSuffix(String filename) {
        AssertUtils.notNull(filename);
        return filename.substring(filename.lastIndexOf("."));
    }

    public static List<URL> listFileAsURL(String dirPath) {
        File dir = new File(dirPath);
        if (dir.exists() && dir.isDirectory()) {
            File[] files = dir.listFiles();
            if (files != null && files.length > 0) {
                List<URL> urls = new ArrayList<>();
                for (File f : files) {
                    try {
                        urls.add(f.toURI().toURL());
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                }
                return urls;
            }
        }
        return Collections.emptyList();
    }

    public static boolean exists(Serializable file) {
        if (file == null) {
            return false;
        }
        if (file instanceof File) {
            return ((File) file).exists();
        }
        return new File(file.toString()).exists();
    }

    public static boolean directoryNotBlank(Serializable file) {
        if (file == null) {
            return false;
        }
        File f = file instanceof File ? (File) file : new File(file.toString());
        String[] list = f.list();
        return f.isDirectory() && list != null && list.length > 0;
    }

    public static boolean equals(File file1, File file2) {
        try {
            return equalsInternal(file1, file2);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static boolean equalsInternal(File file1, File file2) throws IOException {
        if (file1 == null || file2 == null) {
            return false;
        }
        if (!file1.exists() || !file2.exists()) {
            return false;
        }
        if (file1.getAbsolutePath().equals(file2.getAbsolutePath())) {
            return true;
        }
        BufferedInputStream first = new BufferedInputStream(new FileInputStream(file1));
        BufferedInputStream second = new BufferedInputStream(new FileInputStream(file2));
        if (first.available() != second.available()) {
            Utils.close(first, second);
            return false;
        }
        while (true) {
            int firRead = first.read();
            int secRead = second.read();
            if (firRead != secRead) {
                Utils.close(first, second);
                return false;
            }
            if (firRead == -1) {
                Utils.close(first, second);
                return true;
            }
        }
    }

    public static void readInputStream(InputStream in, byte[] array) throws IOException {
        try (InputStream input = in) {
            int toRead = array.length;
            int off = 0;
            while (toRead > 0) {
                int ret = input.read(array, off, toRead);
                if (ret < 0) {
                    throw new IOException("Bad inputStream, premature EOF");
                }
                toRead -= ret;
                off += ret;
            }
        }
    }

    public static String readFile(String filename) throws IOException {
        return SafePathUtils.readConfigFile(filename);
    }

    public static String readFile(File file) throws IOException {
        return readFile(file.getPath());
    }

    public static void writeFile(String content, File file) throws IOException {
        java.nio.channels.WritableByteChannel channel =
            Channels.newChannel(Files.newOutputStream(file.toPath()));
        ByteBuffer buffer = ByteBuffer.wrap(content.getBytes(StandardCharsets.UTF_8));
        channel.write(buffer);
        Utils.close(channel);
    }

    public static byte[] readEndOfFile(File file, long maxSize) throws IOException {
        try (RandomAccessFile raFile = new RandomAccessFile(file, "r")) {
            long readSize = maxSize;
            if (raFile.length() > maxSize) {
                raFile.seek(raFile.length() - maxSize);
            } else if (raFile.length() < maxSize) {
                readSize = raFile.length();
            }
            byte[] fileContent = new byte[(int) readSize];
            raFile.read(fileContent);
            return fileContent;
        }
    }

    public static byte[] readFileFromOffset(File file, long startOffset, long maxSize) throws IOException {
        if (file.length() < startOffset) {
            throw new IllegalArgumentException(
                "The startOffset "
                    + startOffset
                    + " is great than the file length "
                    + file.length());
        }
        try (RandomAccessFile raFile = new RandomAccessFile(file, "r")) {
            long readSize = Math.min(maxSize, file.length() - startOffset);
            raFile.seek(startOffset);
            byte[] fileContent = new byte[(int) readSize];
            raFile.read(fileContent);
            return fileContent;
        }
    }

    public static String tailOf(String path, int offset, int limit) {
        try {
            File file = new File(path);
            if (file.exists() && file.isFile()) {
                return Files.lines(Paths.get(path)).skip(offset).limit(limit).collect(Collectors.joining("\r\n"));
            }
            return null;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static String readString(File file) throws IOException {
        if (file == null || !file.isFile()) {
            throw new IllegalArgumentException("file must be a normal file");
        }
        FileReader reader = new FileReader(file);
        Scanner scanner = new Scanner(reader);
        StringBuilder buffer = new StringBuilder();
        if (scanner.hasNextLine()) {
            buffer.append(scanner.nextLine());
        }
        while (scanner.hasNextLine()) {
            buffer.append("\r\n");
            buffer.append(scanner.nextLine());
        }
        Utils.close(scanner, reader);
        return buffer.toString();
    }
}
